pkg/webhook/hook.go

```diff
@@ -0,0 +1,218 @@
+	hook := &admissionregistration.ValidatingWebhookConfiguration{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "pxc-validation-webhook",
+		},
+		Webhooks: []admissionregistration.ValidatingWebhook{
+			{
+				AdmissionReviewVersions: []string{"v1", "v1beta1"},
+				Name:                    "validationwebhook.pxc.percona.com",
+				ClientConfig: admissionregistration.WebhookClientConfig{
+					Service: &admissionregistration.ServiceReference{
+						Namespace: h.namespace,
+						Name:      "percona-xtradb-cluster-operator",
+						Path:      &hookPath,
+					},
+					CABundle: h.caBunlde,
+				},
+				SideEffects:   &sideEffects,
+				FailurePolicy: &failPolicy,
+				Rules: []admissionregistration.RuleWithOperations{
+					{
+						Rule: admissionregistration.Rule{
+							APIGroups:   []string{"pxc.percona.com"},
+							APIVersions: []string{"*"},
+							Resources:   []string{"*/*"},
+						},
+						Operations: []admissionregistration.OperationType{"CREATE", "UPDATE"},
+					},
+				},
+			},
+		},
+	}
+	err := h.cl.Create(context.TODO(), hook)
+	if err != nil {
+		if k8serrors.IsAlreadyExists(err) {
+			err = h.cl.Delete(context.TODO(), hook)
+			if err != nil {
+				return err
+			}
+			return h.cl.Create(context.TODO(), hook)
+		}
+		return err
+	}
+	return nil
+}
+
+// SetupWebhook prepares certificates for webhook and
+// create ValidatingWebhookConfiguration k8s object
+func SetupWebhook(mgr manager.Manager) error {
+	err := admissionregistration.AddToScheme(mgr.GetScheme())
+	if err != nil {
+		return errors.Wrap(err, "add admissionregistration to scheme")
+	}
+
+	nsBytes, err := ioutil.ReadFile("/var/run/secrets/kubernetes.io/serviceaccount/namespace")
+	if err != nil {
+		return errors.Wrap(err, "read namespace name from file")
+	}
+
+	ca, err := setupCertificates(mgr.GetAPIReader(), string(nsBytes))
+	if err != nil {
+		return errors.Wrap(err, "prepare hook tls certs")
+	}
+
+	h := &hook{cl: mgr.GetClient(), caBunlde: ca, namespace: string(nsBytes)}
+
+	srv := mgr.GetWebhookServer()
+	srv.Port = 9443
+	srv.Register(hookPath, h)
+
+	err = mgr.Add(h)
+	if err != nil {
+		return errors.Wrap(err, "add webhook creator to manager")
+	}
+
+	return nil
+}
+
+func setupCertificates(cl client.Reader, namespace string) ([]byte, error) {
+	certSecret := &corev1.Secret{}
+	err := cl.Get(context.TODO(), types.NamespacedName{
+		Namespace: namespace,
+		Name:      "pxc-webhook-ssl",
+	}, certSecret)
+	if err != nil && !k8serrors.IsNotFound(err) {
+		return nil, err
+	}
+	var ca, crt, key []byte
+	if err != nil {
+		ca, crt, key, err = issueCertificates(namespace)
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		ca, crt, key = certSecret.Data["ca.crt"], certSecret.Data["tls.crt"], certSecret.Data["tls.key"]
+	}
+	return ca, writeCerts(crt, key)
+}
+
+func issueCertificates(namespace string) ([]byte, []byte, []byte, error) {
+	ca, crt, key, err := pxctls.Issue([]string{"percona-xtradb-cluster-operator." + namespace + ".svc"})
```

Comment:
hm, is `.svc` suffix needed?
```suggestion
	ca, crt, key, err := pxctls.Issue([]string{"percona-xtradb-cluster-operator." + namespace})
```

Reply:
Yea, without it it's not working